File: fs/src/main/scala/farjs/fs/popups/FoldersHistoryPopup.scala
package farjs.fs.popups

import farjs.filelist.history.{HistoryKind, HistoryProvider}
import farjs.fs.FSFoldersHistory
import farjs.ui.popup._
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js

object FoldersHistoryPopup extends FunctionComponent[FoldersHistoryPopupProps] {

  private[popups] var listPopup: ReactClass = ListPopup

  private val foldersHistoryKind = FSFoldersHistory.asInstanceOf[js.Dynamic].foldersHistoryKind.asInstanceOf[HistoryKind]
  
  protected def render(compProps: Props): ReactElement = {
    val historyProvider = HistoryProvider.useHistoryProvider()
    val (maybeItems, setItems) = useState(Option.empty[js.Array[String]])
    val props = compProps.plain

    useLayoutEffect({ () =>
      for {
        foldersHistory <- historyProvider.get(foldersHistoryKind).toFuture
        items <- foldersHistory.getAll().toFuture
      } yield {
        setItems(Some(items.map(_.item)))
      }
      ()
    }, Nil)

    maybeItems.map { items =>
      <(listPopup)(^.plain := ListPopupProps(
        title = "Folders history",
        items = items,
        onAction = { index =>
          props.onChangeDir(items(index))
        },
        onClose = props.onClose,
        selected = math.max(items.length - 1, 0)
      ))()
    }.orNull
  }
}

File: fs/popups/FoldersHistoryPopup.mjs
import React, { useLayoutEffect, useState } from "react";
import ListPopup from "@farjs/ui/popup/ListPopup.mjs";
import HistoryProvider from "@farjs/filelist/history/HistoryProvider.mjs";
import FSFoldersHistory from "../FSFoldersHistory.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  onChangeDir(dir: string): void;
 *  onClose(): void;
 * }} FoldersHistoryPopupProps
 */

/**
 * @param {FoldersHistoryPopupProps} props
 */
const FoldersHistoryPopup = (props) => {
  const { listPopup } = FoldersHistoryPopup;

  const historyProvider = HistoryProvider.useHistoryProvider();
  const [maybeItems, setItems] = useState(
    /** @type {readonly string[] | undefined} */ (undefined)
  );

  const readHistory = async () => {
    const foldersHistory = await historyProvider.get(
      FSFoldersHistory.foldersHistoryKind
    );
    const items = await foldersHistory.getAll();
    setItems(items.map((_) => _.item));
  };

  useLayoutEffect(() => {
    readHistory();
  }, []);

  if (maybeItems !== undefined) {
    return h(listPopup, {
      title: "Folders history",
      items: maybeItems,
      onAction: (index) => {
        props.onChangeDir(maybeItems[index]);
      },
      onClose: props.onClose,
      selected: Math.max(maybeItems.length - 1, 0),
    });
  }

  return null;
};

FoldersHistoryPopup.displayName = "FoldersHistoryPopup";
FoldersHistoryPopup.listPopup = ListPopup;

export default FoldersHistoryPopup;
